lib/semantics/check-omp-structure.cc

```diff
@@ -238,12 +206,51 @@ void OmpStructureChecker::Enter(const parser::OmpBlockDirective &x) {
         OmpClause::IF, OmpClause::NUM_THREADS, OmpClause::PROC_BIND};
     SetContextAllowedOnce(allowedOnce);
   } break;
+  // 2.7.3 single-clause -> private-clause |
+  //                        firstprivate-clause
+  case parser::OmpBlockDirective::Directive::Single:
+    SetContextDirectiveEnum(OmpDirective::SINGLE);
+    SetContextAllowed({OmpClause::PRIVATE, OmpClause::FIRSTPRIVATE});
+    break;
+  // 2.7.4 workshare (no clauses are allowed)
+  case parser::OmpBlockDirective::Directive::Workshare:
+    SetContextDirectiveEnum(OmpDirective::WORKSHARE);
+    break;
   default:
     // TODO others
     break;
   }
 }
 
+void OmpStructureChecker::Enter(const parser::OmpEndBlockDirective &x) {
+  const auto &dir{std::get<parser::OmpBlockDirective>(x.t)};
+  SaveContextDirectiveSource();
+  SetContextDirectiveSource(dir.source);
```

Comment:
You may want to simply pop and push a new `OmpContext` here.
Rational:
-   the `allowedClauses`/ `allowedOnceClausesand` seen clauses from the `OmpEndBlockDirective` are still set here, and this can lead to weird thing in the `default` like accepting:
```
real :: a
!$omp parallel
a =1.
!$omp end parallel copyin(a)
end
```
-  I have not looked in detail, but I am a bit surprised that there is a need to save/restore the source location of the `OmpBeginBlockDirective` here given the `OmpContext` stack is basically pop just after restoring it because `Leave(const parser::OmpEndBlockDirective &)` is directly followed by `Leave(const parser::OpenMPBlockConstruct &)`.

Reply:
Thanks... For the first item, I removed the pushing-new-`OmpContext` in the latest commit on Tim's review because there is no "real" context that needs to be pushed (although it was handy). For the second item, I thought about it too, I wasn't sure whether the source location info might be useful in the `Leave` of `OpenMPBlockConstruct`. I will fix both of them.

Reply:
@tskeith I have committed the change for clearing the context before processing `OmpEndBlockDirective`. Thanks for the advice.